plugin/parser/differ/mysql/differ.go

```diff
@@ -139,33 +145,95 @@ func (*SchemaDiffer) SchemaDiff(oldStmt, newStmt string) (string, error) {
 			}
 
 			if len(alterTableDropConstraintSpecs) > 0 {
-				diff = append(diff, &ast.AlterTableStmt{
-					Table: &ast.TableName{
-						Name: model.NewCIStr(tableName),
+				drop = append(drop, []ast.Node{
+					&ast.AlterTableStmt{
+						Table: &ast.TableName{
+							Name: model.NewCIStr(tableName),
+						},
+						Specs: alterTableAddConstraintSpecs,
 					},
-					Specs: alterTableAddConstraintSpecs,
 				})
 			}
 		default:
 		}
 	}
 
-	return deparse(diff, format.DefaultRestoreFlags|format.RestoreStringWithoutCharset)
+	return deparse(add, inplaceUpdate, nonInplaceDrop, nonInplaceAdd, drop, format.DefaultRestoreFlags|format.RestoreStringWithoutCharset)
 }
 
-func deparse(nodes []ast.Node, flag format.RestoreFlags) (string, error) {
+func deparse(add []ast.Node, inplaceUpdate []ast.Node, nonInplaceDrop [][]ast.Node, nonInplaceAdd []ast.Node, drop [][]ast.Node, flag format.RestoreFlags) (string, error) {
 	var buf bytes.Buffer
-	for _, node := range nodes {
+	// We should following the right order to avoid break the dependency:
+	// Additions for new nodes.
+	// Updates for in-place node updates.
+	// Deletions for destructive (none in-place) node updates (in reverse order).
+	// Additions for destructive node updates.
+	// Deletions for deleted nodes (in reverse order).
+	for _, node := range add {
 		if err := node.Restore(format.NewRestoreCtx(flag, &buf)); err != nil {
 			return "", err
 		}
 		if _, err := buf.Write([]byte(";\n")); err != nil {
 			return "", err
 		}
 	}
+	for _, node := range inplaceUpdate {
+		if err := node.Restore(format.NewRestoreCtx(flag, &buf)); err != nil {
+			return "", err
+		}
+		if _, err := buf.Write([]byte(";\n")); err != nil {
+			return "", err
+		}
+	}
+	reNonInplaceDropNodes := reverse2D(nonInplaceDrop)
+	for _, nodes := range reNonInplaceDropNodes {
+		for _, node := range nodes {
+			if err := node.Restore(format.NewRestoreCtx(flag, &buf)); err != nil {
+				return "", err
+			}
+			if _, err := buf.Write([]byte(";\n")); err != nil {
+				return "", err
+			}
+		}
+	}
+	for _, node := range nonInplaceAdd {
+		if err := node.Restore(format.NewRestoreCtx(flag, &buf)); err != nil {
+			return "", err
+		}
+		if _, err := buf.Write([]byte(";\n")); err != nil {
+			return "", err
+		}
+	}
+	reDropNodes := reverse2D(drop)
+	for _, nodes := range reDropNodes {
+		for _, node := range nodes {
+			if err := node.Restore(format.NewRestoreCtx(flag, &buf)); err != nil {
+				return "", err
+			}
+			if _, err := buf.Write([]byte(";\n")); err != nil {
+				return "", err
+			}
+		}
+	}
 	return buf.String(), nil
 }
 
+func reverse2D(nodes [][]ast.Node) [][]ast.Node {
```

Comment:
This function needs a better name.

Reply:
reverseNodes & reverseNodesLists

reverseList & reverseMatrix

Reply:
We inline `reverse2D` and `reverse` manually.